### PR TITLE
Implementation reset password function

### PR DESCRIPTION
**Pull request by RuslanKamenev**

### CODE REVIEW

File: backend/app/Actions/Auth/PasswordChangeAction.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+use App\Exceptions\User\InvalidTokenException;
+use App\Http\Requests\Api\Auth\PasswordChangeRequest;
+use App\Models\PasswordResets;
+use App\Models\User;
+
+final class PasswordChangeAction
+{
+    public function execute(PasswordChangeRequest $request): PasswordChangeResponse
+    {
+        if (!PasswordResets::where('token', $request->get('token'))) {
```
Comment (RomanMahola) [inline backend/app/Actions/Auth/PasswordChangeAction.php:16]: Provide into repository

File: backend/app/Actions/Auth/PasswordChangeAction.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            throw new InvalidTokenException();
+        }
+
+        $resetData = PasswordResets::where('token', $request->get('token'));
+        if($resetData->get()->first()->toArray()['email'] !== $request->get('email')) {
```
Comment (RomanMahola) [inline backend/app/Actions/Auth/PasswordChangeAction.php:21]: Identation

File: backend/app/Actions/Auth/PasswordChangeAction.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            throw new InvalidTokenException();
+        }
+
+        if(User::where('email', $request->get('email'))->update(array('password' => $request->get('password')))) {
```
[the remark this answers is not in the document]
Comment (RomanMahola) [reply backend/app/Actions/Auth/PasswordChangeAction.php:25]: array [
  'password' => ''code"
]

File: backend/app/Console/Kernel.php [hunk 4]
```diff
@@ -15,7 +15,7 @@ class Kernel extends ConsoleKernel
      */
     protected function schedule(Schedule $schedule)
     {
-        // $schedule->command('inspire')->hourly();
+        $schedule->command('auth:clear-resets')->hourly();
```
Comment (RomanMahola) [inline backend/app/Console/Kernel.php:18]: valdiate 19:59

File: backend/app/Http/Controllers/Api/Auth/ForgotPasswordController.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+use App\Http\Requests\Api\Auth\PasswordChangeRequest;
+use Illuminate\Http\Request;
+
+class ForgotPasswordController extends Controller
+{
+    public function passwordResetRequest(ForgotPasswordRequest $request, ForgotPasswordAction $action)
+    {
+        $action->execute($request);
+
+        return response()->json([], 204);
```
Comment (RomanMahola) [inline backend/app/Http/Controllers/Api/Auth/ForgotPasswordController.php:18]: use consts

File: backend/app/Http/Requests/Api/Auth/ForgotPasswordRequest.php [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<?php
+
```
Comment (RomanMahola) [inline backend/app/Http/Requests/Api/Auth/ForgotPasswordRequest.php:2]: declare

File: backend/app/Http/Requests/Api/Auth/PasswordChangeRequest.php [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+use Illuminate\Foundation\Http\FormRequest;
+
+class PasswordChangeRequest extends FormRequest
+{
+    public function rules()
+    {
+        return [
+            "email" => 'required|email',
+            "password" => 'required',
```
Comment (RomanMahola) [inline backend/app/Http/Requests/Api/Auth/PasswordChangeRequest.php:13]: confirmed

File: backend/app/Http/Requests/Api/Auth/PasswordChangeRequest.php [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+            "token" => 'required'
```
Comment (RomanMahola) [inline backend/app/Http/Requests/Api/Auth/PasswordChangeRequest.php:14]: string

File: backend/app/Mail/ForgotPasswordEmail.php [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+use Illuminate\Bus\Queueable;
+use Illuminate\Contracts\Queue\ShouldQueue;
+use Illuminate\Mail\Mailable;
+use Illuminate\Queue\SerializesModels;
+
+class ForgotPasswordEmail extends Mailable
+{
+    use Queueable, SerializesModels;
+
+    /**
```
Comment (RomanMahola) [inline backend/app/Mail/ForgotPasswordEmail.php:14]: delete comments

File: backend/app/Mail/ForgotPasswordEmail.php [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+     * Create a new message instance.
+     *
+     * @return void
+     */
+    public function __construct(public array $token)
```
Comment (RomanMahola) [inline backend/app/Mail/ForgotPasswordEmail.php:19]: empty constructor

File: backend/app/Mail/ForgotPasswordEmail.php [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        //
+    }
+
+    /**
+     * Build the message.
```
Comment (RomanMahola) [inline backend/app/Mail/ForgotPasswordEmail.php:25]: comments

File: backend/database/factories/RoleFactory.php [hunk 12]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace Database\Factories;
+
+use Illuminate\Database\Eloquent\Factories\Factory;
+
+class RoleFactory extends Factory
+{
+    /**
```
Comment (RomanMahola) [inline backend/database/factories/RoleFactory.php:9]: comments

File: backend/resources/views/email/forgot_password.blade.php [hunk 13]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <meta charset="UTF-8">
+    <title>Reset your password</title>
+</head>
+    <body>
+    <h4>Reset password</h4>
+    We got a request to reset your password.
+    <a href="{{ route('password.reset-view', $token['value']) }}">Reset password link</a>
```
Comment (RomanMahola) [inline backend/resources/views/email/forgot_password.blade.php:10]: Check

**Review by RomanMahola — CHANGES_REQUESTED** [review]

**Comment by RuslanKamenev** [comment]
Remove comments, add const and declaration type.

File: backend/app/Http/Controllers/Api/Auth/ForgotPasswordController.php [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+use App\Actions\Auth\ForgotPasswordAction;
+use App\Actions\Auth\PasswordChangeAction;
+use App\Http\Controllers\Controller;
+use App\Http\Requests\Api\Auth\ForgotPasswordRequest;
+use App\Http\Requests\Api\Auth\PasswordChangeRequest;
+use Illuminate\Http\JsonResponse;
+
+class ForgotPasswordController extends Controller
+{
+    const RESPONSE_STATUS_NO_CONTENT = 204;
```
Comment (RuslanKamenev) [inline backend/app/Http/Controllers/Api/Auth/ForgotPasswordController.php:16]: Remove const

File: backend/app/Mail/ForgotPasswordEmail.php [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+use Illuminate\Bus\Queueable;
+use Illuminate\Contracts\Queue\ShouldQueue;
+use Illuminate\Mail\Mailable;
+use Illuminate\Queue\SerializesModels;
+
+class ForgotPasswordEmail extends Mailable
+{
+    use Queueable, SerializesModels;
+    public function __construct(public string $token)
```
Comment (RuslanKamenev) [inline backend/app/Mail/ForgotPasswordEmail.php:13]: Change string to private

File: backend/app/Repositories/PasswordReset/PasswordResetRepository.php [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+namespace App\Repositories\PasswordReset;
+
+use App\Models\PasswordResets;
+use App\Repositories\BaseRepository;
+
+final class PasswordResetRepository extends BaseRepository implements PasswordResetRepositoryInterface
+{
+    public function getByToken(string $token): ?PasswordResets
+    {
+        return PasswordResets::firstWhere('token', $token);
```
Comment (RuslanKamenev) [inline backend/app/Repositories/PasswordReset/PasswordResetRepository.php:14]: Change to findOrFail()

File: backend/database/factories/RoleFactory.php [hunk 17]
```diff
@@ -0,0 +1,15 @@
+<?php
+
```
Comment (RuslanKamenev) [inline backend/database/factories/RoleFactory.php:2]: Change to seed

File: backend/database/seeders/DatabaseSeeder.php [hunk 18]
```diff
@@ -13,6 +14,15 @@ class DatabaseSeeder extends Seeder
      */
     public function run()
     {
+        /**
```
Comment (RuslanKamenev) [inline backend/database/seeders/DatabaseSeeder.php:17]: Remove 17-25

File: backend/resources/views/auth/reset_password.blade.php [hunk 19]
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
```
Comment (RuslanKamenev) [inline backend/resources/views/auth/reset_password.blade.php:1]: Remove file

File: backend/resources/views/errors/404.blade.php [hunk 20]
```diff
@@ -0,0 +1 @@
+<h2>{{ $exception->getMessage() }}</h2>
```
Comment (RuslanKamenev) [inline backend/resources/views/errors/404.blade.php:1]: Remove file

File: backend/resources/views/forgot.blade.php [hunk 21]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
```
Comment (RuslanKamenev) [inline backend/resources/views/forgot.blade.php:1]: Remove file

**Review by RomanMahola — APPROVED** [review]

**Pull request merged**